config/emacs/config.org
#+TITLE: Emacs Configuration
#+PROPERTY: header-args:elisp :tangle ./init.el :mkdirp yes
#+HTML_HEAD: <link rel="stylesheet" type="text/css" href="../css/org.css" />
#+HTML_LINK_UP: index.html
#+HTML_LINK_HOME: ../index.html
#+EXPORT_FILE_NAME: emacs-config

#+begin_src elisp
  ;; init.el -*- lexical-binding: t; -*-

  (add-to-list 'load-path (expand-file-name "./lisp" user-emacs-directory))
#+end_src

* Basic stuff

** Identity
#+begin_src elisp
  (setq user-full-name "nuxsh"
        user-mail-address "nuxshed@gmail.com")
#+end_src

Change the defaults because some of them suck
#+begin_src elisp
  (setq-default require-final-newline t
                vc-follow-symlinks)

  (setq undo-limit 80000000            ; moar undo
        auto-save-default t            ; who knows what could happen?
        truncate-string-ellipsis "…")  ; prettier ellipsis

  (setq use-short-answers t)
#+end_src

UI debloating
#+begin_src elisp
  (setq inhibit-startup-echo-area-message t
        inhibit-startup-message t)
  (tool-bar-mode -1)     ; Disable the toolbar
  (tooltip-mode -1)      ; Disable tooltips
  (menu-bar-mode -1)     ; Disable the menu bar
  (scroll-bar-mode -1)   ; Disable the scrollbar
#+end_src

optimizations
#+begin_src elisp
  (setq frame-inhibit-implied-resize t
        initial-major-mode 'fundamental-mode
        file-name-handler-alist nil)
#+end_src

  Set custom file and load it
#+begin_src elisp
  (setq custom-file "~/.emacs.d/custom.el")
  (load custom-file)
#+end_src

Some useful minor modes
#+begin_src elisp
  (recentf-mode) ;; recent files
  (electric-pair-mode) ;; autopairs
#+end_src

* Packages

Setting up use-package
#+begin_src elisp
  ;; init package sources
  (require 'package)
  (setq package-archives '(("gnu" . "http://elpa.gnu.org/packages/")
			   ("nongnu" . "https://elpa.nongnu.org/nongnu/")
			   ("melpa" . "http://melpa.org/packages/")))
  (package-initialize)
  (unless package-archive-contents
    (package-refresh-contents))

  ;; Initialize use-package on non-Linux platforms
  (unless (package-installed-p 'use-package)
    (package-install 'use-package))

  (require 'use-package)
  (setq use-package-always-ensure t
	package-quickstart t)
#+end_src

** gcmh
#+begin_src elisp
  (use-package gcmh
    :init
    (gcmh-mode 1))
#+end_src

** Evil
#+begin_src elisp
  (use-package evil
    :init
    (setq evil-want-integration t)
    (setq evil-want-keybinding nil)
    :config
    (evil-mode 1))

  (use-package evil-collection
    :after evil
    :config
    (evil-collection-init))

  (setq evil-cross-lines t
        evil-move-beyond-eol t
        evil-symbol-word-search t
        evil-want-Y-yank-to-eol t
        evil-cross-lines t)
#+end_src

** Leader
#+begin_src elisp
  (use-package evil-leader
    :init
    (global-evil-leader-mode)
    (evil-leader/set-leader "<SPC>")
    (evil-leader/set-key
      ;; General
      ".f" 'consult-isearch
      ".q" 'delete-frame
      ".e" 'eval-region
      ;; Files
      "fr" 'consult-recent-file
      "fb" 'consult-bookmark
      "ff" 'find-file
      "fd" 'dired
      ;; Org
      "oa" 'org-agenda
      "fh" 'consult-org-heading
      ;; Open
      "om" 'mu4e
      "os" 'eshell
      ;; Notes
      "no" 'deft
      "nf" 'deft-find-file
      "nn" 'deft-new-file-named
      ;; Bufffers
      "bd" 'kill-current-buffer
      "bb" 'consult-buffer
      "bx" 'switch-to-scratch
      "bi" 'ibuffer
      ;; Windows
      "wv" 'split-window-right
      "wh" 'split-window-below
      "wt" 'window-split-toggle
      "ws" 'ace-window
      ;; Help
      "hh" 'help
      "hk" 'describe-key
      "hv" 'describe-variable
      "hF" 'describe-function
      "hf" 'describe-face
      "hs" 'describe-symbol
      "hm" 'describe-mode))
#+end_src

** Consult
#+begin_src elisp
  (use-package consult)
#+end_src

** Vertico
#+begin_src elisp
  (use-package vertico
    :init
    (vertico-mode)
    (setq vertico-cycle t))
#+end_src

Hotfuzz adds fuzzy finding
#+begin_src elisp
  (use-package hotfuzz
    :init
    (setq completion-styles '(hotfuzz)))
#+end_src

Marginalia adds descriptions and stuff to your minibuffer
#+begin_src elisp
  (use-package marginalia
    :init (marginalia-mode))
#+end_src

** Company-mode
#+begin_src elisp
  (use-package company
    :bind (:map company-active-map
                ("<tab>" . company-select-next)))

  (use-package company-statistics
    :hook (company-mode . company-statistics-mode))

  (use-package company-quickhelp
    :hook (company-mode . company-quickhelp-mode))

  (use-package company-box
    :hook (company-mode . company-box-mode)
    :config
    (setq company-box-scrollbar nil))

  (add-hook 'after-init-hook 'global-company-mode)
#+end_src

** Vterm
#+begin_src elisp
  (use-package vterm
    :commands (vterm vterm-other-window))
#+end_src

** Which-key
#+begin_src elisp
  (use-package which-key
    :config (which-key-mode)
    (which-key-setup-side-window-bottom)
    (setq which-key-idle-delay 0.1))
#+end_src

** Snippets
#+begin_src elisp
  (use-package yasnippet
    :hook (org-mode . yas-global-mode))
#+end_src

** Languages
Some major modes to improve language support
#+begin_src elisp
  (use-package clojure-mode)
  (use-package nix-mode)
  (use-package lua-mode)
#+end_src

*** Lispy
#+begin_src elisp
  (use-package lispy
    :hook ((emacs-lisp-mode clojure-mode) . lispy-mode))
#+end_src

*** Cider
#+begin_src elisp
  (use-package cider :defer t)
#+end_src

** Flycheck
#+begin_src elisp
  (use-package flycheck
    :ensure t
    :config
    (setq flycheck-elisp-load-path 'inherit)
    (setq flycheck-idle-change-delay 1.0)
    (setq-local flycheck-elisp-initialize-packages t)
    (setq-local flycheck-elisp-package-user-dir package-user-dir)
    (setq-default flycheck-disabled-checkers '(elisp-checkdoc)))

    (add-hook 'prog-mode-hook 'flycheck-mode)
#+end_src

** Rainbow-mode
Visualize color values.
#+begin_src elisp
  (use-package rainbow-mode
    :defer t)
#+end_src

* Builtins

** Dired
#+begin_src elisp
  (with-eval-after-load 'dired
    (setq dired-dwim-target t
          dired-listing-switches "-Alh"
          dired-use-ls-dired t
          dired-omit-files "\\`[.]?#\\|\\`[.][.]?\\|\\`[.].*\\'"
          dired-always-read-filesystem t
          dired-create-destination-dirs 'ask
          dired-hide-details-hide-symlink-targets nil
          dired-isearch-filenames 'dwim)
    (define-key dired-mode-map (kbd "^") (lambda () (interactive) (find-alternate-file ".."))))
  (add-hook 'dired-mode-hook 'dired-hide-details-mode)
  (add-hook 'dired-mode-hook 'dired-omit-mode)

  (use-package all-the-icons-dired)
  (add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
  (setq all-the-icons-dired-monochrome 't)
#+end_src

** Ibuffer

Modify the default ibuffer formats
#+begin_src elisp
  (setq ibuffer-formats
        '((mark modified read-only locked " "
                (name 20 20 :left :elide)
                " "
                (size-h 11 -1 :right)
                " "
                (mode 16 16 :left :elide))
          (mark " "
                (name 16 -1)
                " " filename)))
#+end_src

Set filter groups
#+begin_src elisp
  (setq ibuffer-show-empty-filter-groups nil)
  (setq ibuffer-saved-filter-groups
	'(("main"
	   ("modified" (and
			(modified . t)
			(visiting-file . t)))
	   ("term" (or
		    (mode . vterm-mode)
		    (mode . eshell-mode)
		    (mode . term-mode)
		    (mode . shell-mode)))
	   ("planning" (or
			(name . "^\\*Calendar\\*$")
			(name . "^agenda")
			(mode . org-agenda-mode)))
	   ("img" (mode . image-mode))
	   ("config" (filename . "/dotfiles/"))
	   ("blog" (filename . "/projects/site/"))
	   ("code" (or (filename . "/projects/")
		       (filename . "/code/")))
	   ("notes" ( filename . "/notes/"))
	   ("org" (mode . org-mode))
	   ("dired" (mode . dired-mode))
	   ("help" (or (name . "\*Help\*")
		       (name . "\*Apropos\*")
		       (name . "\*info\*")
		       (mode . help-mode)))
	   ("internal" (name . "^\*.*$"))
	   ("other" (name . "^.*$")))))

  (add-hook 'ibuffer-mode-hook
	    (lambda ()
	      (ibuffer-auto-mode 1)
	      (ibuffer-switch-to-saved-filter-groups "main")))
#+end_src

Add icons to ibuffer
#+begin_src elisp
  (use-package all-the-icons-ibuffer
    :ensure t
    :init (all-the-icons-ibuffer-mode 1))
#+end_src

* UI

** Font
#+begin_src elisp
  (set-face-attribute 'default nil :font "Cartograph CF 10")
  (set-face-attribute 'fixed-pitch nil :font "Cartograph CF 10")
  (set-face-attribute 'variable-pitch nil :font "Commissioner 10")
#+end_src

** Icons
#+begin_src elisp
  (use-package all-the-icons :defer t)
#+end_src

** Theme
#+begin_src elisp
  (use-package doom-themes
    :config
    (load-theme 'doom-kurai t))
#+end_src

** Layout
#+begin_src elisp
  (fringe-mode 10)
  (add-to-list 'default-frame-alist '(internal-border-width . 24))
#+end_src

** Modeline
#+begin_src elisp
  (defun mode-line-render (left right)
    "Return a string of `window-width' length.
  Containing LEFT, and RIGHT aligned respectively."
    (let ((available-width
	    (- (window-width)
	      (+ (length (format-mode-line left))
		(length (format-mode-line right))))))
      (append left
	(list (format (format "%%%ds" available-width) ""))
	right)))

  (setq-default mode-line-format
    '((:eval (mode-line-render
	       '((:eval (propertize " %b" 'face `(:slant italic)))
		 (:eval (if (and buffer-file-name (buffer-modified-p))
			    (propertize "*" 'face `(:inherit face-faded))))
		 (:eval (if (buffer-narrowed-p)
			   (propertize "-" 'face `(:inherit face-faded)))))
	       '("%p %l:%c "
		 (:eval (propertize " %m" 'face 'font-lock-string-face)))))))

#+end_src

** Splash
#+begin_src elisp
  (require 'splash)
  (splash-screen)
#+end_src

** Smooth scrolling
#+begin_src elisp
  (use-package good-scroll
    :config
    (good-scroll-mode 1))
#+end_src

* Window Stuff

Configure how a particular buffer appears.
#+begin_src elisp
  (setq display-buffer-alist
	'(
	  ("*Help*"
	   (display-buffer-in-side-window)
	   (window-width 0.25)
	   (side . right)
	   (slot . 0))
	  ("\\*e?shell.*"
	   (display-buffer-in-side-window)
	   (window-height . 0.35)
	   (side . bottom)
	   (slot . -1))))
#+end_src

** Functions

Fun functions.
#+begin_src elisp
  (defun window-split-toggle ()
    "Toggle between horizontal and vertical split."
    (interactive)
    (if (> (length (window-list)) 2)
	(error "wont work lol")
      (let ((func (if (window-full-height-p)
		      #'split-window-vertically
		    #'split-window-horizontally)))
	(delete-other-windows)
	(funcall func)
	(save-selected-window
	  (other-window 1)
	  (switch-to-buffer (other-buffer))))))

  ;; got this off of some random gist from 2012
  (defun toggle-split-view ()
    "Toggle between split window and single window."
    (interactive)
    (if (not (window-minibuffer-p (selected-window)))
	(progn
	  (if (< 1 (count-windows))
	      (progn
		(window-configuration-to-register ?u)
		(delete-other-windows))
	    (jump-to-register ?u))))
    (my-iswitchb-close))

  (define-key global-map (kbd "C-`") 'toggle-split-view)

  ;; iswitchb hasn't been included in emacs 24.4, so i have no idea what this does
  (defun my-iswitchb-close()
   "Open iswitchb or, if in minibuffer go to next match."
   (interactive)
   (if (window-minibuffer-p (selected-window))
      (keyboard-escape-quit)))
#+end_src

** Ace-window
Quite useful plugin for switching between buffers
#+begin_src elisp
  (use-package ace-window
    :defer t)
#+end_src

* Mail
#+begin_src elisp
  (require 'mu4e)

  (setq mu4e-maildir (expand-file-name "~/.mail/"))

  (setq mu4e-drafts-folder "/Gmail/[Gmail]/Drafts")
  (setq mu4e-sent-folder   "/Gmail/[Gmail]/Sent Mail")
  (setq mu4e-trash-folder  "/Gmail/[Gmail]/Trash")

  (setq mu4e-get-mail-command "mbsync -a"
        mu4e-compose-signature-auto-include nil
        mu4e-compose-format-flowed t)

  (setq
   user-mail-address "nuxshed@gmail.com"
   user-full-name  "nuxsh")

  (setq mu4e-view-show-images t)

  (setq smtpmail-smtp-server "smtp.gmail.com"
        user-mail-address "nuxshed@gmail.com"
        smtpmail-smtp-user "nuxshed"
        smtpmail-smtp-service 587)

  (setq smtpmail-auth-credentials (expand-file-name "~/.authinfo"))
#+end_src

* Org-mode
#+begin_src elisp
  (use-package org-contrib)
  (use-package org-bullets
    :after org
    :hook
    (org-mode . org-bullets-mode))

  (setq org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●"))
  (setq org-agenda-files '("~/org/agenda.org"))
  
  (setq org-hide-leading-stars t
	org-hide-emphasis-markers t
	org-hide-macro-markers t
	org-link-descriptive t)

  (add-hook 'org-agenda-mode-hook
	    (lambda ()
	      (local-set-key (kbd "q") 'org-agenda-exit)))
  (use-package htmlize)
  (add-hook 'org-mode-hook (lambda ()
			     (toggle-truncate-lines)
			     (flyspell-mode t)
			     (electric-indent-local-mode -1)))
  (setq org-src-window-setup 'current-window)
#+end_src

** Deft

Deft is nice for note-taking
#+begin_src elisp
  (use-package deft
    :config
    (setq deft-directory "~/notes"
          deft-default-extension "org"
          deft-extensions '("txt" "md" "org")
          deft-use-filter-string-for-filename t))

  (add-hook 'deft-mode-hook
            (lambda ()
              (define-key evil-normal-state-local-map (kbd "n") 'deft-new-file-named)
              (define-key evil-normal-state-local-map (kbd "q") 'quit-window)
              (define-key evil-normal-state-local-map (kbd "f") 'deft-find-file)))
#+end_src

** Publishing
#+begin_src elisp
  (require 'publish)
#+end_src

** Prettify symbols
Make org-mode prettier

#+begin_src elisp
  (defun org/prettify-set ()
    (interactive)
    (setq prettify-symbols-alist
	  '(("#+begin_src" . ">")
	    ("#+BEGIN_SRC" . ">")
	    ("#+end_src" . "<")
	    ("#+END_SRC" . "<")
	    ("#+begin_example" . "")
	    ("#+BEGIN_EXAMPLE" . "")
	    ("#+end_example" . "")
	    ("#+END_EXAMPLE" . "")
	    ("#+results:" . "")
	    ("#+RESULTS:" . "")
	    ("#+begin_quote" . "❝")
	    ("#+BEGIN_QUOTE" . "❝")
	    ("#+end_quote" . "❞")
	    ("#+END_QUOTE" . "❞"))))
  (add-hook 'org-mode-hook 'org/prettify-set)

  (defun prog/prettify-set ()
    (interactive)
    (setq prettify-symbols-alist
	  '(("lambda" . "λ")
	    ("->" . "→")
	    ("<-" . "←")
	    ("<=" . "≤")
	    (">=" . "≥")
	    ("!=" . "≠")
	    ("~=" . "≃")
	    ("=~" . "≃"))))
  (add-hook 'prog-mode-hook 'prog/prettify-set)

  (global-prettify-symbols-mode)
#+end_src

the end
#+begin_src elisp
  ;; init.el ends here
#+end_src


emacs/.doom.d/config.el
;;; config.el -*- lexical-binding: t; -*-

(remove-hook 'org-mode-hook #'+literate-enable-recompile-h)

(setq user-full-name "nuxsh"
      user-mail-address "nuxshed@gmail.com")

(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to lose work
      scroll-margin 2                             ; Having a little margin is nice
      truncate-string-ellipsis "…"                ; Unicode ellispis are nicer than "...", and also save /precious/ space
      display-line-numbers-type t
      which-key-idle-delay 0.3                    ; Show key binding help quicker
      which-key-idle-secondary-delay 0
      shell-file-name "/bin/dash"                 ; Use dash as default shell for running term which is faster
      vterm-always-compile-module t               ; Compile the vterm-module when needed without asking
      vterm-shell "/bin/zsh")                     ; Use zsh in vterm)

(global-subword-mode 1) ; navigate through Camel Case words

(setq doom-font (font-spec :family "FiraCode Nerd Font" :size 11)
      doom-big-font (font-spec :family "FiraCode Nerd Font" :size 16)
      doom-variable-pitch-font (font-spec :family "Montserrat" :size 12))
      doom-serif-font (font-spec :family "Montserrat" :size 12)

(setq doom-theme 'doom-one)
(setq doom-one-padded-modeline t)
(setq doom-themes-treemacs-theme "doom-atom")

(after! projectile
  (setq projectile-project-root-files-bottom-up '("package.json" ".projectile" ".project" ".git")
        projectile-ignored-projects '("~/.emacs.d/")
        projectile-project-search-path '("~/projects"))
  (defun projectile-ignored-project-function (filepath)
    "Return t if FILEPATH is within any of `projectile-ignored-projects'"
    (or (mapcar (lambda (p) (s-starts-with-p p filepath)) projectile-ignored-projects))))

(after! popup
  (set-popup-rule! "^\\*Flycheck errors\\*$" :side 'bottom :size 0.2 :select t))

(after! flycheck
  (setq flycheck-check-syntax-automatically '(mode-enabled save new-line idle-change)))

(after! (flycheck lsp-mode)
  (add-hook 'lsp-after-initialize-hook (lambda()
                                        (flycheck-add-next-checker 'lsp '(warning . javascript-eslint)))))

;; There's a weird bug where fringe-modes < 8 dont show the fringes
(after! git-gutter
  (fringe-mode 8)
  (after! git-gutter-fringe
    (fringe-mode 8))
  (setq +vc-gutter-diff-unsaved-buffer t))

(after! ibuffer
  (set-popup-rule! "^\\*Ibuffer\\*$" :side 'bottom :size 0.4 :select t :ignore nil))

(setq doom-themes-treemacs-theme "doom-colors")
(after! treemacs
  (setq +treemacs-git-mode 'extended)
  (treemacs-follow-mode t))
(after! rainbow-mode
  (setq rainbow-html-colors-major-mode-list '(html-mode css-mode php-mode nxml-mode xml-mode typescript-mode javascript-mode)))

(set-frame-parameter nil 'internal-border-width 15)
(setq-default left-margin-width 2)
(setq-default right-margin-width 2)

;;modeline (icons, config, battery)
(display-time-mode 1)                              ;Enable time in the mode-line
(display-battery-mode 1)                           ;display the battery
(setq doom-modeline-major-mode-icon t)             ;Show major mode name
(setq doom-modeline-enable-word-count t)           ;Show word count
(setq doom-modeline-modal-icon t)                  ;Show vim mode icon
(setq inhibit-compacting-font-caches t)            ;Don't compact font caches in gc
(setq doom-modeline-lsp t)

(defun doom-modeline-conditional-buffer-encoding ()
  "We expect the encoding to be LF UTF-8, so only show the modeline when this is not the case"
  (setq-local doom-modeline-buffer-encoding
              (unless (and (memq (plist-get (coding-system-plist buffer-file-coding-system) :category)
                                 '(coding-category-undecided coding-category-utf-8))
                           (not (memq (coding-system-eol-type buffer-file-coding-system) '(1 2))))
                t)))
(add-hook 'after-change-major-mode-hook #'doom-modeline-conditional-buffer-encoding) ;;remove encoding

(after! centaur-tabs
  (centaur-tabs-mode -1)
  (setq centaur-tabs-height 30
        centaur-tabs-set-icons t
        centaur-tabs-gray-out-icons 'buffer)
  (centaur-tabs-group-by-projectile-project)

  (+popup-window-p) ; needed to prevent recursive auto-loading of popup

  ;; Automatically turn off tabs in popups
  (defun +fl/hide-tabs-in-popup ()
    (if (+popup-window-p)
        (centaur-tabs-local-mode)
      (centaur-tabs-local-mode 0)))
  (add-hook! 'buffer-list-update-hook '+fl/hide-tabs-in-popup))

(use-package! lsp-ui
  :hook (lsp-mode . lsp-ui-mode)
  :config
  (setq lsp-ui-sideline-enable nil; not anymore useful than flycheck
        lsp-ui-doc-enable nil
        lsp-enable-symbol-highlighting nil))

(cl-defmacro lsp-org-babel-enable (lang)
  "Support LANG in org source code block."
  (setq centaur-lsp 'lsp-mode)
  (cl-check-type lang stringp)
  (let* ((edit-pre (intern (format "org-babel-edit-prep:%s" lang)))
         (intern-pre (intern (format "lsp--%s" (symbol-name edit-pre)))))
    `(progn
       (defun ,intern-pre (info)
         (let ((file-name (->> info caddr (alist-get :file))))
           (unless file-name
             (setq file-name (make-temp-file "babel-lsp-")))
           (setq buffer-file-name file-name)
           (lsp-deferred)))
       (put ',intern-pre 'function-documentation
            (format "Enable lsp-mode in the buffer of org source block (%s)."
                    (upcase ,lang)))
       (if (fboundp ',edit-pre)
           (advice-add ',edit-pre :after ',intern-pre)
         (progn
           (defun ,edit-pre (info)
             (,intern-pre info))
           (put ',edit-pre 'function-documentation
                (format "Prepare local buffer environment for org source block (%s)."
                        (upcase ,lang))))))))
(defvar org-babel-lang-list
  '("python" "ipython" "bash" "sh" "rust" "lua"))
(dolist (lang org-babel-lang-list)
  (eval `(lsp-org-babel-enable ,lang)))

(set-docsets! 'python-mode "Python 3")
(set-docsets! 'lua-mode "Lua")
(set-docsets! 'emacs-lisp-mode "Emacs Lisp")
(setq +lookup-open-url-fn #'+lookup-xwidget-webkit-open-url-fn)

(setq org-directory "~/org/"
      org-ellipsis "  "
      org-journal-file-type 'weekly
      org-use-property-inheritance t
      org-log-done 'time
      org-hide-emphasis-markers t
      org-enforce-todo-dependencies t
      org-enforce-todo-checkbox-dependencies t
      org-log-into-drawer t
      org-log-state-notes-into-drawer t
      org-log-repeat 'time
      org-todo-repeat-to-state "TODO")

(after! org
  (setq org-tags-column -80
        org-agenda-sticky nil))

(advice-add 'org-refile :after 'org-save-all-org-buffers)
(advice-add 'org-gcal-fetch :after 'org-save-all-org-buffers)

(after! org
  (setq org-tags-column -80)
  (appendq! +ligatures-extra-symbols
            `(:checkbox      ""
              :doing         ""
              :checkedbox    ""
              :list_property "∷"))
  (set-ligatures! 'org-mode
    :merge t
    :checkbox      "[ ]"
    :doing         "[-]"
    :checkedbox    "[X]"
    :list_property "::"))

(setq deft-directory "~/notes")
(setq deft-extensions '("txt" "tex" "org" "md"))
(setq deft-recursive t)

(defvar mixed-pitch-modes '(org-mode LaTeX-mode markdown-mode gfm-mode Info-mode)
  "Modes that `mixed-pitch-mode' should be enabled in, but only after UI initialisation.")
(defun init-mixed-pitch-h ()
  "Hook `mixed-pitch-mode' into each mode in `mixed-pitch-modes'.
Also immediately enables `mixed-pitch-modes' if currently in one of the modes."
  (when (memq major-mode mixed-pitch-modes)
    (mixed-pitch-mode 1))
  (dolist (hook mixed-pitch-modes)
    (add-hook (intern (concat (symbol-name hook) "-hook")) #'mixed-pitch-mode)))
(add-hook 'doom-init-ui-hook #'init-mixed-pitch-h)

(after! web-mode
  (add-hook 'web-mode-hook #'flycheck-mode)

  (setq web-mode-markup-indent-offset 2 ;; Indentation
        web-mode-code-indent-offset 2
        web-mode-enable-auto-quoting nil ;; disbale adding "" after an =
        web-mode-auto-close-style 2))

(set-email-account! "gmail"
  '((mu4e-sent-folder       . "/gmail/\[Gmail\]/Sent Mail")
    (mu4e-drafts-folder     . "/gmail/\[Gmail\]/Drafts")
    (mu4e-trash-folder      . "/gmail/\[Gmail\]/Trash")
    (mu4e-refile-folder     . "/gmail/\[Gmail\]/All Mail")
    (smtpmail-smtp-user     . "nuxshed@gmail.com")
    (user-mail-address      . "nuxshed@gmail.com")    ;; only needed for mu < 1.4
    (mu4e-compose-signature . "---\nnuxsh"))
  t)

(setq +mu4e-gmail-accounts '(("nuxshed@gmail.com" . "/gmail")))

(require 'smtpmail)
(setq message-send-mail-function 'smtpmail-send-it
   starttls-use-gnutls t
   smtpmail-starttls-credentials '(("smtp.gmail.com" 587 nil nil))
   smtpmail-auth-credentials
     '(("smtp.gmail.com" 587 "nuxshed@gmail.com" nil))
   smtpmail-default-smtp-server "smtp.gmail.com"
   smtpmail-smtp-server "smtp.gmail.com"
   smtpmail-smtp-service 587)
